solidity/contracts/libs/EnumerableMOfNSet.sol

```diff
@@ -0,0 +1,232 @@
+     * @param _set The set to check membership in
+     * @param _value The address being checked for set membership
+     * @return Whether or not the set contains `_value`
+     */
+    function contains(AddressSet storage _set, address _value)
+        internal
+        view
+        returns (bool)
+    {
+        return _set.addresses.contains(_value);
+    }
+
+    /**
+     * @notice Returns number of values in the set
+     * @param _set The set to return the length of
+     * @return The number of values in the set
+     */
+    function length(AddressSet storage _set) internal view returns (uint256) {
+        return _set.addresses.length();
+    }
+
+    /**
+     * @notice Returns whether or two sets are identical
+     * @param _set One of the two sets to check for equality
+     * @param _threshold The threshold of the second set
+     * @param _values The addresses in the second set
+     * @return Whether or not the two sets are identical
+     */
+    function matches(
+        AddressSet storage _set,
+        uint8 _threshold,
+        bytes calldata _values
+    ) internal view returns (bool) {
+        bytes32 _commitment = keccak256(abi.encodePacked(_threshold, _values));
+        return _commitment == _set.commitment;
+    }
+
+    /**
+     * @notice Returns all addresses in the set and its threshold
+     * @param _set The set whose addresses and threshold are returned
+     * @return All addresses in the set and its threshold
+     */
+    function valuesAndThreshold(AddressSet storage _set)
+        internal
+        view
+        returns (address[] memory, uint8)
+    {
+        return (values(_set), _set.threshold);
+    }
+
+    // ============ Internal Functions ============
+
+    /**
+     * @notice Adds an address to the set without updating the commitment
+     * @param _set The set to add to
+     * @param _value The address to add to the set
+     */
+    function _add(AddressSet storage _set, address _value) private {
+        require(_value != address(0), "zero address");
+        require(_set.addresses.add(_value), "contained");
+    }
+
+    /**
+     * @notice Removes an address from the set without updating the commitment
+     * @param _set The set to add to
+     * @param _value The address to remove from the set
+     */
+    function _remove(AddressSet storage _set, address _value) private {
+        require(_set.addresses.remove(_value), "!contained");
+        require(_set.addresses.length() >= _set.threshold, "reduce threshold");
+    }
+
+    /**
+     * @notice Updates the set commitment to match its content
+     * @param _set The set whose commitment should be updated
+     * @return The updated commitment
+     */
+    function _updateCommitment(AddressSet storage _set)
+        private
+        returns (bytes32)
+    {
+        bytes32 _commitment = _computeCommitment(
+            _set.threshold,
+            _set.addresses.values()
```

Comment:
from the values() natspec
 
 > WARNING: This operation will copy the entire storage to memory, which can be quite expensive. This is designed to mostly be used by view accessors that are queried without any gas fees. Developers should keep in mind that this function has an unbounded cost, and using it as part of a state-changing function may render the function uncallable if the set grows to a point where copying to memory consumes too much gas to fit in a block.

Reply:
What would you suggest as an alternative?